File: python/character_replacement.py
class Solution(object):
    # O(n) time | O(26) space
    def characterReplacement(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        l = 0
        maxFreq = 0
        counts = [0] * 26
        res = 0

        for r, c in enumerate(s):
            counts[ord(c) - ord('A')] += 1
            maxFreq = max(maxFreq, counts[ord(c) - ord('A')])

            while (r - l + 1) - maxFreq > k:
                t = s[l]
                counts[ord(t) - ord('A')] -= 1
                l += 1
            
            res = max(res, r - l + 1)

        return res
    
solution = Solution()
s = "AABABBA"
k = 1
print(solution.characterReplacement(s, k))

File: src/main/java/july_2025/CharacterReplacement.java
package july_2025;

import java.util.HashMap;
import java.util.Map;

public class CharacterReplacement {

    public static void main(String[] args) {
//        String s = "AABABBA";
//        int k = 1;

        String s = "ABAB";
        int k = 2;

        CharacterReplacement characterReplacement = new CharacterReplacement();
        int result = characterReplacement.characterReplacement(s, k);
        System.out.println(result);
    }

    // O(n) time | O(26) space
    public int characterReplacement(String s, int k) {
        int[] freqMap = new int[26];
        int start = 0;
        int res = 0;
        int maxFreq = 0;
        for (int i = 0; i < s.length(); i++) {
            char c1 = s.charAt(i);
            freqMap[c1 - 'A']++;
            maxFreq = Math.max(maxFreq, freqMap[c1 - 'A']);
            while (i - start - maxFreq >= k) {
                char c2 = s.charAt(start);
                freqMap[c2 - 'A']--;
                maxFreq = Math.max(maxFreq, freqMap[c2 - 'A']);
                start++;
            }
            res = Math.max(res, i - start + 1);
        }
        return res;
    }

}
